src/sdk/pynni/nni/nas/pytorch/cream/trainer.py

```diff
@@ -0,0 +1,308 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import os
+import logging
+from copy import deepcopy
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as dist
+from nni.nas.pytorch.trainer import Trainer
+from nni.nas.pytorch.utils import AverageMeterGroup
+
+logger = logging.getLogger(__name__)
+
+
+class CreamSupernetTrainer(Trainer):
+    """
+    This trainer trains a supernet that can be used for evolution search.
+
+    Parameters
+    ----------
+    model : nn.Module
+        Model with mutables.
+    mutator : Mutator
+        A mutator object that has been initialized with the model.
+    loss : callable
+        Called with logits and targets. Returns a loss tensor.
+    metrics : callable
+        Returns a dict that maps metrics keys to metrics data.
+    optimizer : Optimizer
+        Optimizer that optimizes the model.
+    num_epochs : int
+        Number of epochs of training.
+    train_loader : iterable
+        Data loader of training. Raise ``StopIteration`` when one epoch is exhausted.
+    dataset_valid : iterable
+        Data loader of validation. Raise ``StopIteration`` when one epoch is exhausted.
+    batch_size : int
+        Batch size.
+    workers: int
+        Number of threads for data preprocessing. Not used for this trainer. Maybe removed in future.
+    device : torch.device
+        Device object. Either ``torch.device("cuda")`` or ``torch.device("cpu")``. When ``None``, trainer will
+        automatic detects GPU and selects GPU first.
+    log_frequency : int
+        Number of mini-batches to log metrics.
+    callbacks : list of Callback
+        Callbacks to plug into the trainer. See Callbacks.
+    """
+
+    def __init__(self, model, loss,
+                 optimizer, num_epochs, train_loader, valid_loader,
+                 mutator=None, batch_size=64, log_frequency=None,
+                 est=None, meta_sta_epoch=20, update_iter=200, slices=2, pool_size=10,
```

Comment:
`est`: estimation, used for estimate model flops through a lookup table; type: obj (has two dicts). how to produce `est`, current layer with current resolution;